### ISSUE TITLE
TF2 Memory Leak when fitting multiple models in Docker

### ISSUE THREAD

**Issue by markodjordjic** [tensorflow/tensorflow#43627]
When fitting multiple models within same tuning process, the consumption of RAM grows until all of the memory is consumed, which ultimately leads to failure of the complete training process.

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution: Linux Ubuntu 16.04
- TensorFlow installed from (source or binary): docker pull tensorflow/tensorflow:2.0.0-py3-jupyter
- TensorFlow version (use command below): tensorflow:2.0.0  and every one up!
- Python version: 3.6 

**Describe the current behavior**
When fitting multiple keras models RAM is consumed until there is no more free RAM and training process fails.

**Describe the expected behavior**
tf.keras.backend.clear_session() should reset the memory consumption after each tuning iteration.

**Standalone code to reproduce the issue**
from tensorflow import keras
import psutil

for i in range(0, 200):
    keras.backend.clear_session()
    a = keras.layers.Input(shape=(32,))
    b = keras.layers.Dense(32)(a)
    model = keras.Model(inputs=a, outputs=b)
    print('memory usesd: ' + str(psutil.virtual_memory().used // 1e6))

**Comment by bhack** [comment]
Can you try with TF 2.3.1?

**Comment by markodjordjic** [comment]
@bhack that TF image is not available on Docker Hub.

**Comment by bhack** [comment]
2.3.0 is ok too..

**Comment by markodjordjic** [comment]
@bhack yes pls take a look at: https://github.com/tensorflow/tensorflow/issues/31312

**Comment by yhliang2018** [comment]
#31312 is closed already. Let's use this issue thread to discuss the memory leak problem.

@markodjordjic Could you verify [this comment](https://github.com/tensorflow/tensorflow/issues/43627#issuecomment-700225043) first?

**Comment by markodjordjic** [comment]
@bhack yes TF2.3.0 does not help

**Comment by bhack** [comment]
It seems to me that it isn't  growing in colab with 2.3.0

**Comment by markodjordjic** [comment]
@bhack can we please acknowledge that there is a problem. Numerous issues have been opened and there is no progress.

**Comment by bhack** [comment]
On TF 1.3.0 on Colab
```python
from tensorflow import keras
from statistics import mean
import psutil
mem_stats=[]
for i in range(0, 10001):
  keras.backend.clear_session()
  a = keras.layers.Input(shape=(32,))
  b = keras.layers.Dense(32)(a)
  model = keras.Model(inputs=a, outputs=b)
  mem_stats.append(psutil.virtual_memory().used)
  if (i % 1000 == 0):
    print('memory -> step %5d Max %d Min %d Mean %d ' % (i, 
                                                         max(mem_stats)/(1024), 
                                                         min(mem_stats)/(1024), 
                                                         mean(mem_stats)/(1024)
                                                         )
    )
    mem_stats.clear()
```

```
memory -> step     0 Max 1262812 Min 1262812 Mean 1262812 
memory -> step  1000 Max 1265760 Min 1261540 Mean 1262152 
memory -> step  2000 Max 1265484 Min 1261376 Mean 1261946 
memory -> step  3000 Max 1270308 Min 1261772 Mean 1264956 
memory -> step  4000 Max 1269436 Min 1264424 Mean 1265338 
memory -> step  5000 Max 1268056 Min 1264468 Mean 1265161 
memory -> step  6000 Max 1269956 Min 1264492 Mean 1265200 
memory -> step  7000 Max 1268564 Min 1264476 Mean 1265354 
memory -> step  8000 Max 1268628 Min 1264376 Mean 1265071 
memory -> step  9000 Max 1268000 Min 1263448 Mean 1264206 
memory -> step 10000 Max 1267624 Min 1263276 Mean 1263966 
```

**Comment by markodjordjic** [comment]
@bhack yes. On lower versions of TF it works. I had to refactor the code to lower versions.

**Comment by bhack** [comment]
It was just a typo.. is 2.3.0

**Comment by markodjordjic** [comment]
@bhack I have utilized the docker pull tensorflow/tensorflow:2.3.1-gpu-jupyter and I can confirm that the issue is _not_ resolved. Can you please take into account that the training is happening inside docker container on AWS and that I could be that keras.backend.clear_session() cannot release the RAM occupied by tf.keras,model in this setting.

**Comment by bhack** [comment]
Can you run my example unmodified on your  setuo and paste the output here?

**Comment by markodjordjic** [comment]
@bhack here you go. TF docker image tensorflow/tensorflow:2.3.1-gpu-jupyter.

algo-1-7b5eo_1  | 2020-09-29 14:14:14.820908: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
algo-1-7b5eo_1  | memory -> step     0 Max 806536 Min 806536 Mean 806536 
algo-1-7b5eo_1  | memory -> step  1000 Max 813376 Min 806332 Mean 811834 
algo-1-7b5eo_1  | memory -> step  2000 Max 816452 Min 813352 Mean 815939 
algo-1-7b5eo_1  | memory -> step  3000 Max 816444 Min 816436 Mean 816440 
algo-1-7b5eo_1  | memory -> step  4000 Max 816676 Min 816436 Mean 816590 
algo-1-7b5eo_1  | memory -> step  5000 Max 816916 Min 816676 Mean 816773 
algo-1-7b5eo_1  | memory -> step  6000 Max 820372 Min 816784 Mean 817007 
algo-1-7b5eo_1  | memory -> step  7000 Max 817528 Min 817288 Mean 817397 
algo-1-7b5eo_1  | memory -> step  8000 Max 817512 Min 817280 Mean 817384 
algo-1-7b5eo_1  | memory -> step  9000 Max 817504 Min 817376 Mean 817381 
algo-1-7b5eo_1  | memory -> step 10000 Max 817748 Min 817368 Mean 817393 
algo-1-7b5eo_1  | 2020-09-29 14:15:41,387 sagemaker-training-toolkit INFO     Reporting training SUCCESS
tmpri_3aih3_algo-1-7b5eo_1 exited with code 0
Aborting on container exit...
===== Job Complete =====

**Comment by markodjordjic** [comment]
@bhack it revealed the problem.

**Comment by bhack** [comment]
I've try in the same docker image that you have used but on a machine without cuda devices and it doesn't seems to me growing monotonically. At 10k steps is less then 3k steps:
```
memory -> step     0 Max 3873720 Min 3873720 Mean 3873720 
memory -> step  1000 Max 3883720 Min 3873720 Mean 3880353 
memory -> step  2000 Max 3884048 Min 3882772 Mean 3883289 
memory -> step  3000 Max 3885432 Min 3883848 Mean 3884593 
memory -> step  4000 Max 3884192 Min 3883912 Mean 3884009 
memory -> step  5000 Max 3887120 Min 3883696 Mean 3885244 
memory -> step  6000 Max 3885896 Min 3885248 Mean 3885504 
memory -> step  7000 Max 3885656 Min 3884416 Mean 3884744 
memory -> step  8000 Max 3885172 Min 3884440 Mean 3884854 
memory -> step  9000 Max 3885676 Min 3884280 Mean 3884742 
memory -> step 10000 Max 3885228 Min 3884152 Mean 3884441 

```

Can you verify if your problem is on GPU only? Run the same with this before the loop
```
 tf.config.set_visible_devices([], 'GPU')
```

**Comment by markodjordjic** [comment]
@bhack the AWS instance on which I ran the script was CPU only.

**Comment by markodjordjic** [comment]
@bhack any ideas how to circumvent this in CPU only setting? Btw, thanks for your efforts.

**Comment by bhack** [comment]
I cannot reproduce this. 
I've runned the script with mprof for 50k steps in the Docker container:
`mprof run --include-children --multiprocess tf_memory_test.py`

![memory](https://user-images.githubusercontent.com/1710528/94676566-d5267600-031b-11eb-8de4-1eebde3c403c.png)

**Comment by markodjordjic** [comment]
@bhack fit method of the keras.model object is the main problem. If fit (and predict) methods are added there is a memory leak.

**Comment by bhack** [comment]
So you have another minimal example with fit/predict to reproduce this?

**Comment by markodjordjic** [comment]
@bhack I will make one tomorrow. I have also a work-around for this issue. I will also post it tomorrow. It has been a long day here. Thank you.

**Comment by markodjordjic** [comment]
@bhack here is an example with fit method.
for i in range(0, 10001):
  keras.backend.clear_session()
  a = keras.layers.Input(shape=(2,))
  b = keras.layers.Dense(32)(a)
  output = keras.layers.Dense(1)(b)
  model = keras.Model(inputs=a, outputs=output)
  model.compile(
    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
    loss='mse',
    metrics=None,
  )
  model.fit(
    x=x,
    y=y,
    batch_size=128,
    epochs=2,
    verbose=0,
    callbacks=None,
    shuffle=False    
  )
  print('memory uses: ' + str(psutil.virtual_memory().used // 1e6))

**Comment by bhack** [comment]
This code has missing x,y can you share something runnable?

**Comment by markodjordjic** [comment]
@bhack how did that happen. Here is an update.
import numpy as np
from tensorflow import keras
from statistics import mean
import psutil

# Generate data.
samples = 10000
mean = [0, 0]
cov = [[1, 0], [0, 100]]  # diagonal covariance
x = np.random.multivariate_normal(mean, cov, samples)
y = np.random.normal(0, 1, (samples, 1))
for i in range(0, 10001):
  keras.backend.clear_session()
  a = keras.layers.Input(shape=(2,))
  b = keras.layers.Dense(32)(a)
  output = keras.layers.Dense(1)(b)
  model = keras.Model(inputs=a, outputs=output)
  model.compile(
    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
    loss='mse',
    metrics=None,
  )
  model.fit(
    x=x,
    y=y,
    batch_size=128,
    epochs=2,
    verbose=0,
    callbacks=None,
    shuffle=False    
  )
  print('Memory used: ' + str(psutil.virtual_memory().used // 1e6))

**Comment by bhack** [comment]
Can you edit a final version that we can just copy and run? 
Please put it formatted in a code block as I've done with my examples.

**Comment by markodjordjic** [comment]
`import numpy as np
from tensorflow import keras
from statistics import mean
import psutil

samples = 10000
mean = [0, 0]
cov = [[1, 0], [0, 100]]  # diagonal covariance
x = np.random.multivariate_normal(mean, cov, samples)
y = np.random.normal(0, 1, (samples, 1))
for i in range(0, 10001):
  keras.backend.clear_session()
  a = keras.layers.Input(shape=(2,))
  b = keras.layers.Dense(32)(a)
  output = keras.layers.Dense(1)(b)
  model = keras.Model(inputs=a, outputs=output)
  model.compile(
    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
    loss='mse',
    metrics=None,
  )
  model.fit(
    x=x,
    y=y,
    batch_size=128,
    epochs=2,
    verbose=0,
    callbacks=None,
    shuffle=False    
  )
  print('Memory used: ' + str(psutil.virtual_memory().used // 1e6))`

**Comment by markodjordjic** [comment]
@bhack there you go

**Comment by markodjordjic** [comment]
@bhack @yhliang2018 there appears to be more of this: https://github.com/tensorflow/tensorflow/issues/43702

**Comment by bersbersbers** [comment]
Hi, yes, #43702 seems very related, also because that code over there fails to reproduce the issue when there is no `model.fit`.

In addition, #43702 has a complete example with all imports and data that reproduces the issue on a CPU-only system and on Colab, as well as on Linux and Windows. Finally, that example reports process memory consumption rather than system memory consumption. In summary, I feel the example in #43702 may be better suited to repro the issue.

**Comment by markodjordjic** [comment]
@bersbersbers @bhack funny enough docker pull tensorflow/tensorflow:nightly-gpu-jupyter seems to help. I need to run a more comprehensive training process, to confirm, but on the example with fit method, the memory consumption stays fairly managable. Just to confirm this is AWS linux setup, on a ml.t3.medium instance and the training is done within the docker container mentioned above.

**Comment by markodjordjic** [comment]
@bhack @yhliang2018 Hi. Thank you for your efforts. I would like to confirm that the issue still persists on AWS regardless of docker pull tensorflow/tensorflow:nightly-gpu-jupyter. The workaround is to implement custom training process and circumvent the usage of keras.model.fit method. Here is an example which can be customized to particular needs:

https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch

Is there any way that the existence of this issue can be acknowledged and addressed in future releases of the TF?

**Comment by bhack** [comment]
Have you tried to compile with eager like in https://github.com/tensorflow/tensorflow/issues/43702#issuecomment-702681590

**Comment by markodjordjic** [comment]
@bhack Thank you for your reply. I have run a minimalistic example, and I can confirm that `run_eagerly=True` reduces the memory leak, however the problem still exists. After 30 cv folds the consumption of RAM has increased from 891 MB to 940 MB. Therefore, approximately 50 MB per single training cycle. In case of 300 models, this will lead to consumption of approximately 15 Gb over the complete training process. Also, it is unclear if this will scale up with the increase of the size of the training set (larger training set, larger memory leak).
Therefore, the problem is ameliorated, however it still exists and it will exhibit it self in more complex training processes, involving more models, or maybe larger training sets.

**Comment by bhack** [comment]
Can you graph with `mprun` like in https://github.com/tensorflow/tensorflow/issues/43702#issuecomment-703510259

**Comment by jvishnuvardhan** [comment]
@markodjordjic Is this still an issue for you? Can you please check with recent TF versions (`TF2.7` anmd `tf-nightly`). Thanks!
